### DISCUSSION TITLE
[ Enhancement ] MariaDB Sequence support

### DISCUSSION THREAD

**Discussion by davidducos** [mydumper/mydumper#931] [Ideas]
@LinuxJedi wrote to me:

```
Hi David,

I have most of the code done for sequences support, but there is a
problem that needs solving first.

Sequences are similar to views in how they exist. A table can use a
sequence in it as a column. So all sequences need to be restored before
table schemas, otherwise table schemas will error.

The only way right now I can see of doing this is to do the "create
databases", then the sequences, then the tables in the main
`process_stream_queue`.

Do you have any other ideas for ways around this?
```

And I answered this:

```
mmmmm... I'm not familiar with sequences, so I reviewed 
what was the current behaviour and I found the first problem:
MariaDB [empty_db4]> show create table test;
| test  | CREATE TABLE `test` (
  `id` int(11) NOT NULL DEFAULT nextval(`empty_db2`.`s`),
  `val` int(11) DEFAULT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_general_ci |

That is a table that I imported with myloader using as 
default value nextval. I didn't create the table with the 
schema name, this means that a relative table creation is 
being transfort to absolute. I think that this is something 
that MariaDB needs to change, because that is not something 
that happens with triggers. What I mean with this, is that if you 
create the table like this:
show create table test;
CREATE TABLE `test` (
  `id` int(11) NOT NULL DEFAULT nextval(`empty_db2`.`s`),
  `val` int(11) DEFAULT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_general_ci

Then it is ok that the SHOW CREATE TABLE to show that. But I created like this:
show create table test;
| test  | CREATE TABLE `test` (
  `id` int(11) NOT NULL DEFAULT nextval(`s`),
  `val` int(11) DEFAULT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_general_ci
I'm testing in MariaDB 10.5.

Alternatively, we can replace  "`schema`" and "`schema_in_-B`" but it is awful 
from my point of view. Or fail if we are using -B and we found nexval.

If that can be solved, then we need to identify the tables that are USING 
sequence to be created after OR we can try to create them and if it fails
 we can reenqueue which is something that we are going to implement as we have 
issues on MySQL 8. 
```

**Comment by LinuxJedi** [comment]
This is a case where the cat is already out of the bag as it were, even if we change the behaviour in MariaDB now, it doesn't retrospectively change things for current MariaDB users.

What we can do on the mydumper side is a simple regex during dump (or even myloader restore time?) to find `nextval` and alike, match to see if the current schema is referenced and remove the schema name from it if it is.

**Reply by davidducos** [reply]
you are right, we can't do anything for old versions...

If we change anything on mydumper, it will need to detect that nextval is being used and delete the ``` `*\`. ``` inside parentheses. This is quite easy and we are doing it for skip definer. Actually, we can do it on myloader too... OK, for now, we can add an option like: --remove-schema-on-nextval, add the proper function on common.c and add the logic to mydumper and myloader. In the future, I think that we need to find a better way to implement this changes in the statements.

**Comment by LinuxJedi** [comment]
Note that I found out mariadb-dump (MariaDB's mysqldump) is also affected by the issue you found. I've created https://jira.mariadb.org/browse/MDEV-30155 for this.

**Comment by davidducos** [comment]
you are right, we can't do anything for old versions...

If we change anything on mydumper, it will need to detect that nextval is being used and delete the ``` `*\`. ``` inside parentheses. This is quite easy and we are doing it for skip definer. Actually, we can do it on myloader too... OK, for now, we can add an option like: --remove-schema-on-nextval, add the proper function on common.c and add the logic to mydumper and myloader. In the future, I think that we need to find a better way to implement this changes in the statements.

**Comment by LinuxJedi** [comment]
OK, a filter for this has been added to my branch with the option `--remove-schema-on-sequence` (as it could be `nextval()` or `lastval()`).

We still have the original problem though, sequences need to be restored before tables are. Best way I can think of so far is a queue that runs before the stream queue. Open to ideas here.

**Reply by davidducos** [reply]
I think that we could re execute create table statements, re enqueuing the job in a "waiting queue"... but I need to re think this after releasing v0.13.1-2
